### PR TITLE
fix Segment#intersects to compare bound instead of token

### PR DESCRIPTION
**Pull request by jasonstack**

### CODE REVIEW

File: src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java [hunk 1]
```diff
@@ -87,14 +87,15 @@ public boolean intersects(AbstractBounds<PartitionPosition> keyRange)
         if (keyRange instanceof Range && ((Range<?>)keyRange).isWrapAround())
             return keyRange.contains(minKeyBound) || keyRange.contains(maxKeyBound);
 
-        int cmp = keyRange.right.getToken().compareTo(minKey);
+        int cmp = keyRange.right.compareTo(minKeyBound);
         // if right is minimum, it means right is the max token and bigger than maxKey.
         // if right bound is less than minKey, no intersection
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java]:
```suggestion
        // if right bound is less than minKeyBound, no intersection
```

File: src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java [hunk 2]
```diff
@@ -87,14 +87,15 @@ public boolean intersects(AbstractBounds<PartitionPosition> keyRange)
...
         if (!keyRange.right.isMinimum() && (!keyRange.inclusiveRight() && cmp == 0 || cmp < 0))
             return false;
 
+        cmp = keyRange.left.compareTo(maxKeyBound);
         // if left bound is bigger than maxKey, no intersection
-        return keyRange.isStartInclusive() || keyRange.left.getToken().compareTo(maxKey) < 0;
+        return (keyRange.isStartInclusive() || cmp != 0) && cmp <= 0;
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java:92]: With the changes in this method, we no longer need the `minKey` or `maxKey` fields in this class, right?

CC @mike-tr-adamson

File: src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java [hunk 3]
```diff
@@ -87,14 +87,15 @@ public boolean intersects(AbstractBounds<PartitionPosition> keyRange)
...
 
-        int cmp = keyRange.right.getToken().compareTo(minKey);
+        int cmp = keyRange.right.compareTo(minKeyBound);
         // if right is minimum, it means right is the max token and bigger than maxKey.
         // if right bound is less than minKey, no intersection
         if (!keyRange.right.isMinimum() && (!keyRange.inclusiveRight() && cmp == 0 || cmp < 0))
             return false;
 
+        cmp = keyRange.left.compareTo(maxKeyBound);
         // if left bound is bigger than maxKey, no intersection
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java]:
```suggestion
        // if left bound is bigger than maxKeyBound, no intersection
```

**Review by maedhroz — CHANGES_REQUESTED** [review]
@jasonstack Mostly looks good. Just a few little things to clean up as a result of the changes...

File: src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java (hunk 2 above)
Comment (jasonstack) [reply src/java/org/apache/cassandra/index/sai/disk/v1/segment/Segment.java:92]: removed `minKey` and `maxKey`

**Review by maedhroz — APPROVED** [review]

**Pull request merged**